### PR TITLE
refactor: system v2 stream conditions format

### PR DESCRIPTION
**Pull request by jsolaas**
Add priority and stream names
Use names of priority, consumer and stream to set stream conditions

## Why is this pull request needed?

Allow multiple streams (more than inlet, outlet) as input to system, also to avoid using indices for everything.

## What does this pull request change?

stream conditions input for consumer system. And I guess we lied when we said that the new consumer system would be less verbose! I guess it might be less verbose in the worst cases with a lot of temporal models..

### CODE REVIEW

File: src/libecalc/fixtures/cases/consumer_system_v2/data/consumer_system_v2.yaml [hunk 1]
```diff
@@ -85,15 +85,75 @@ INSTALLATIONS:
...
+                  inlet:
+                    RATE:
+                      VALUE: 3000000
+                    PRESSURE:
+                      VALUE: 50
+                    FLUID_DENSITY:
+                      VALUE: 2
+                  outlet:
+                    PRESSURE:
+                      VALUE: 125
```
Comment (jsolaas) [inline src/libecalc/fixtures/cases/consumer_system_v2/data/consumer_system_v2.yaml:156]: B-E-A-utiful

File: src/libecalc/fixtures/cases/consumer_system_v2/data/consumer_system_v2.yaml [hunk 2]
```diff
@@ -85,15 +85,75 @@ INSTALLATIONS:
...
-              - RATES: [ 2000000, 2500000, 3000000 ]
-                INLET_PRESSURE: 50
-                OUTLET_PRESSURE: 125
-                FLUID_DENSITY: 2
+            STREAM_CONDITIONS_PRIORITIES:
+              pri1:
+                pump1:
+                  inlet:
+                    RATE:
+                      VALUE: 4000000
```
Comment (jsolaas) [inline src/libecalc/fixtures/cases/consumer_system_v2/data/consumer_system_v2.yaml:93]:
```
                    RATE:
                      VALUE: 4000000
                      UNIT: SM3/d
```

Make room for units, or condition? Condition might make sense one level up? Condition means "empty stream"?

File: src/libecalc/fixtures/cases/consumer_system_v2/data/consumer_system_v2.yaml [hunk 3]
```diff
@@ -135,13 +195,88 @@ INSTALLATIONS:
...
-            INLET_PRESSURE: 50
-            OUTLET_PRESSURE: 250
-          - RATES: [ "$var.compressor1", 5000000, 5000000 ]
-            INLET_PRESSURE: 50
-            OUTLET_PRESSURE: 125
-          - RATES: [ 1000000, 5000000, 5000000 ]
-            INLET_PRESSURE: 50
-            OUTLET_PRESSURE: 125
+        STREAM_CONDITIONS_PRIORITIES:
+          pri1:
```
Comment (jsolaas) [inline src/libecalc/fixtures/cases/consumer_system_v2/data/consumer_system_v2.yaml:199]: I think we should provide an option to not specify priorities, just stream conditions directly.

We should also figure out how defaults could be set, default for a priority, default for all priorities, default for a compressor? Unit makes sense to set for all stream conditions in most cases I assume.

File: src/libecalc/common/stream.py [hunk 4]
```diff
@@ -15,9 +14,9 @@ class Config:
         alias_generator = to_camel_case
         allow_population_by_field_name = True
 
-    name: Optional[str]
-    rate: TimeSeriesStreamDayRate
-    pressure: TimeSeriesFloat
+    name: str
+    rate: Optional[TimeSeriesStreamDayRate]
+    pressure: Optional[TimeSeriesFloat]
     fluid_density: Optional[TimeSeriesFloat] = None
```
Comment (olelod) [inline src/libecalc/common/stream.py:20]: At some point we need to put composition of fluid model in here, not just the density. Different streams entering the compressor train/system/network should be allowed to have different fluid models.

File: src/libecalc/common/stream.py [hunk 5]
```diff
@@ -15,9 +14,9 @@ class Config:
...
 
     def mix(self, other_stream: "Stream") -> "Stream":
```
Comment (olelod) [inline src/libecalc/common/stream.py:22]: This method will of course become more involved/advanced when we allow self and other_stream to have different compositions.

File: src/libecalc/core/consumers/base/component.py [hunk 6]
```diff
@@ -22,15 +20,8 @@ def evaluate(
...
-    @property
     @abstractmethod
-    def operational_settings(self):
-        ...
-
-    @abstractmethod
-    def get_max_rate(
-        self, operational_settings: Union[CompressorOperationalSettings, PumpOperationalSettings]
-    ) -> List[float]:
+    def get_max_rate(self, inlet_stream: Stream, target_pressure: TimeSeriesFloat) -> List[float]:
```
Comment (olelod) [inline src/libecalc/core/consumers/base/component.py:24]: this one also gets more complicated when/if we allow multiple streams in/out of a consumer/compressor train/system/network. 

`def get_max_rate_per_stream(self, streams: List[Stream]) -> List[float]`

or something. The maximum of one stream given that all others are kept constant?

File: src/libecalc/common/priority_optimizer.py [hunk 7]
```diff
@@ -8,16 +8,19 @@
...
 TResult = TypeVar("TResult")
 TPriority = TypeVar("TPriority")
 
 
 @dataclass
 class PriorityOptimizerResult(Generic[TResult]):
-    priorities_used: TimeSeriesInt
-    priority_results: Dict[datetime, Dict[int, Dict[str, TResult]]]
+    priorities_used: TimeSeriesString
+    priority_results: Dict[datetime, Dict[str, Dict[str, TResult]]]
```
Comment (TeeeJay) [inline src/libecalc/common/priority_optimizer.py:23]:
```suggestion
    priority_results: Dict[datetime, Dict[str, Dict[PriorityID, TResult]]]
```
?

File: src/libecalc/common/priority_optimizer.py [hunk 8]
```diff
@@ -54,15 +57,19 @@ def optimize(
...
         """
         is_valid = TimeSeriesBoolean(timesteps=timesteps, values=[False] * len(timesteps), unit=Unit.NONE)
-        priorities_used = TimeSeriesInt(timesteps=timesteps, values=[0] * len(timesteps), unit=Unit.NONE)
-        priority_results: Dict[datetime, Dict[int, Dict[str, TResult]]] = defaultdict(dict)
+        priorities_used = TimeSeriesString(
+            timesteps=timesteps,
+            values=[list(priorities.keys())[0]] * len(timesteps),
+            unit=Unit.NONE,
+        )
+        priority_results: Dict[datetime, Dict[str, Dict[str, TResult]]] = defaultdict(dict)
```
Comment (TeeeJay) [inline src/libecalc/common/priority_optimizer.py:65]:
```suggestion
        priority_results: Dict[datetime, Dict[str, Dict[PriorityID, TResult]]] = defaultdict(dict)
```
?

File: src/libecalc/common/priority_optimizer.py [hunk 9]
```diff
@@ -54,15 +57,19 @@ def optimize(
 
         """
         is_valid = TimeSeriesBoolean(timesteps=timesteps, values=[False] * len(timesteps), unit=Unit.NONE)
-        priorities_used = TimeSeriesInt(timesteps=timesteps, values=[0] * len(timesteps), unit=Unit.NONE)
-        priority_results: Dict[datetime, Dict[int, Dict[str, TResult]]] = defaultdict(dict)
+        priorities_used = TimeSeriesString(
+            timesteps=timesteps,
+            values=[list(priorities.keys())[0]] * len(timesteps),
```
Comment (TeeeJay) [inline src/libecalc/common/priority_optimizer.py:62]: initialiseres til første pri? best å initialisere til 0 (hvis det betyr ingeting) eller None?

File: src/libecalc/common/priority_optimizer.py [hunk 10]
```diff
@@ -80,15 +87,15 @@ def optimize(
...
                 # Register the valid timesteps as valid and keep track of the operational setting used
                 is_valid[new_valid_indices] = True
-                priorities_used[new_valid_indices] = priority_index
+                priorities_used[new_valid_indices] = priority_name
 
                 if all(is_valid.values):
                     # quit as soon as all time-steps are valid. This means that we do not need to test all settings.
                     break
-                elif priority_index + 1 == len(priorities):
+                elif list(priorities.keys()).index(priority_name) + 1 == len(priorities):
```
Comment (TeeeJay) [inline src/libecalc/common/priority_optimizer.py:95]: funksjon for å hente riktig id? mtp index+1?

File: src/libecalc/common/stream.py [hunk 11]
```diff
@@ -31,16 +30,24 @@ def mix(self, other_stream: "Stream") -> "Stream":
         Returns: The mixed stream
 
         """
+
+        if self.rate is None or other_stream.rate is None:
```
Comment (TeeeJay) [inline src/libecalc/common/stream.py:34]: ta med navn mtp kontekst? den under også

File: src/tests/libecalc/integration/snapshots/test_consumer_system_v2/test_compressor_system_v2_results/consumer_system_v2-consumer_system_v20/consumer_system_v2.json [hunk 12]
```diff
@@ -192,15 +192,15 @@
...
+                                "unit": "kg/m3",
                                 "values": [
                                     2.0,
                                     2.0,
                                     2.0,
                                     2.0
                                 ]
                             },
-                            "name": null,
+                            "name": "Mixed-inlet- crossover-81798f85c9244fbd547076bbaa2d5746-5e69b2c1b61362f59f53c7b2f02ed4d9- crossover-81fd4a36dee423b6bf8143654336d3cd-5e69b2c1b61362f59f53c7b2f02ed4d9",
```
Comment (TeeeJay) [inline src/tests/libecalc/integration/snapshots/test_consumer_system_v2/test_compressor_system_v2_results/consumer_system_v2-consumer_system_v20/consumer_system_v2.json:203]: :O

Comment (TeeeJay) [reply src/tests/libecalc/integration/snapshots/test_consumer_system_v2/test_compressor_system_v2_results/consumer_system_v2-consumer_system_v20/consumer_system_v2.json:203]: litt inkonsistent mtp spaces etter dash

File: src/tests/libecalc/integration/snapshots/test_consumer_system_v2/test_compressor_system_v2_results/consumer_system_v2-consumer_system_v20/consumer_system_v2.json [hunk 13]
```diff
@@ -1409,22 +1339,8 @@
...
-                                "unit": "kg/Sm3",
-                                "values": [
-                                    2.0,
-                                    2.0,
-                                    2.0,
-                                    2.0
-                                ]
-                            },
-                            "name": null,
+                            "fluid_density": null,
```
Comment (TeeeJay) [inline src/tests/libecalc/integration/snapshots/test_consumer_system_v2/test_compressor_system_v2_results/consumer_system_v2-consumer_system_v20/consumer_system_v2.json:1342]: fluid_density fra 2 til null? bug?

**Review by TeeeJay — APPROVED** [review]
noen kommentarer...se over ellers ok :)

File: src/libecalc/common/priority_optimizer.py (hunk 9 above)
Comment (jsolaas) [reply src/libecalc/common/priority_optimizer.py:62]: Antageligvis fordi man må initialisere til en godkjent type, kunne vel satt til siste pri. Er en elif lenger ned som setter til siste pri hvis det er invalid igjen.

File: src/libecalc/common/priority_optimizer.py (hunk 10 above)
Comment (jsolaas) [reply src/libecalc/common/priority_optimizer.py:95]: Kunne vel droppet denne hvis vi initialiserte til siste pri?

**Pull request merged**